From d890c12767b430dc0fe5a4300b891c25deaa5772 Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Thu, 21 Sep 2017 16:58:57 +0800
Subject: [PATCH 2570/5242] MLK-16522-1 usb: cdns3: change cdns3_role_start
 work flow

commit  015b78a1144bf96b7cb65137fe4ad23dc3e4f8bc from
https://source.codeaurora.org/external/imx/linux-imx.git

When it goes to start new role, the interrupt may be occurred before
role_start returns, but at this time, the cdns->role is still the old
role, so the interrupt handler will make mistake.

In this commit, we set desired role before role_start, if the role_start
has failed and the desired role is different with current one, it tries
to back current role.

BuildInfo:
- SCFW 1f59442e, IMX-MKIMAGE fb52c576, ATF
- U-Boot 2017.03-imx_v2017.03+g34be5a2

Acked-by: Li Jun <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/cdns3/core.c |   15 ++++++++++++++-
 drivers/usb/cdns3/core.h |   10 ++--------
 2 files changed, 16 insertions(+), 9 deletions(-)

diff --git a/drivers/usb/cdns3/core.c b/drivers/usb/cdns3/core.c
index b640197..296fcfc 100644
--- a/drivers/usb/cdns3/core.c
+++ b/drivers/usb/cdns3/core.c
@@ -364,15 +364,28 @@ static void cdns3_remove_roles(struct cdns3 *cdns)
 
 static int cdsn3_do_role_switch(struct cdns3 *cdns, enum cdns3_roles role)
 {
+	int ret = 0;
+	enum cdns3_roles current_role;
+
 	dev_dbg(cdns->dev, "current role is %d, switch to %d\n",
 			cdns->role, role);
 
 	if (cdns->role == role)
 		return 0;
 
+	current_role = cdns->role;
 	cdns3_role_stop(cdns);
 	cdns_set_role(cdns, role);
-	return cdns3_role_start(cdns, role);
+	ret = cdns3_role_start(cdns, role);
+	if (ret) {
+		/* Back to current role */
+		dev_err(cdns->dev, "set %d has failed, back to %d\n",
+					role, current_role);
+		cdns_set_role(cdns, current_role);
+		ret = cdns3_role_start(cdns, current_role);
+	}
+
+	return ret;
 }
 
 /**
diff --git a/drivers/usb/cdns3/core.h b/drivers/usb/cdns3/core.h
index 1381e9f..5a37de4 100644
--- a/drivers/usb/cdns3/core.h
+++ b/drivers/usb/cdns3/core.h
@@ -88,16 +88,11 @@ static inline struct cdns3_role_driver *cdns3_role(struct cdns3 *cdns)
 
 static inline int cdns3_role_start(struct cdns3 *cdns, enum cdns3_roles role)
 {
-	int ret;
-
 	if (!cdns->roles[role])
 		return -ENXIO;
 
-	ret = cdns->roles[role]->start(cdns);
-	if (!ret)
-		cdns->role = role;
-
-	return ret;
+	cdns->role = role;
+	return cdns->roles[role]->start(cdns);
 }
 
 static inline void cdns3_role_stop(struct cdns3 *cdns)
@@ -105,7 +100,6 @@ static inline void cdns3_role_stop(struct cdns3 *cdns)
 	enum cdns3_roles role = cdns->role;
 
 	cdns->roles[role]->stop(cdns);
-	cdns->role = CDNS3_ROLE_GADGET;
 }
 
 #endif /* __DRIVERS_USB_CDNS3_CORE_H */
-- 
1.7.9.5

